src/pytorch_lightning/core/module.py

```diff
@@ -172,6 +173,21 @@ def lr_schedulers(self) -> Optional[Union[LRSchedulerTypeUnion, List[LRScheduler
         # multiple schedulers
         return lr_schedulers
 
+    @property
+    def trainer(self) -> "pl.Trainer":
```

Comment:
@awaelchli I'm pretty sure I broke the spawn launcher with this change:


```python
pytest trainer/properties/test_get_model.py::test_get_model_ddp_cpu
E         File "/home/carmocca/git/lightning/src/pytorch_lightning/trainer/trainer.py", line 2474, in training
E           return self.state.stage == RunningStage.TRAINING
E       AttributeError: 'Trainer' object has no attribute 'state'
```

But I'm not seeing the cause. Can you tell what I missed?

Reply:
@carmocca it looks like the trainer holding the loop and the trainer reference inside the fit loop are not the same. If you add
`assert self == self.fit_loop.trainer`
in line 1146 of trainer.py you get the assertion error. On master, this assertion holds true, so it must be something related to the setter/getter for trainer in the loops where it breaks.

Reply:
If I remove the two lines

```py
        if not isinstance(trainer, weakref.ProxyTypes):
            trainer = weakref.proxy(trainer)
```
then the references are the same again and the error does not occur.

Reply:
Looks like the spawn pickling mechanism converts them to different objects, and since they don't share memory, the state is not set on the latter.

I'm going to undo the proxy change for the loops and see if it works for the LM at least